File: Development/volume/graph_generator.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Struct to store an edge
typedef struct {
    long long u;
    long long v;
    int weight;
} Edge;

void generate_weighted_graph(long long vertices, long long edges, const char *filename) {
    if (edges < vertices - 1) {
        fprintf(stderr, "Number of edges is too small to form a connected graph.\n");
        exit(EXIT_FAILURE);
    }
    if (edges > vertices * (vertices - 1) / 2) {
        fprintf(stderr, "Too many edges for the given number of vertices.\n");
        exit(EXIT_FAILURE);
    }

    // Allocate memory for the edge list
    Edge *edge_list = malloc(edges * sizeof(Edge));
    if (!edge_list) {
        fprintf(stderr, "Memory allocation failed.\n");
        exit(EXIT_FAILURE);
    }

    srand(time(NULL));

    // Step 1: Add edges to form a spanning tree (ensure connectivity)
    long long edge_count = 0;
    for (long long i = 1; i < vertices; i++) {
        long long u = i;
        long long v = rand() % i; 
        int weight = (rand() % 100) + 1;
        edge_list[edge_count++] = (Edge){u, v, weight};
    }

    // Step 2: Add remaining edges randomly
    while (edge_count < edges) {
        long long u = rand() % vertices;
        long long v = rand() % vertices;
        if (u != v) { // Ensure no self-loops
            int weight = (rand() % 100) + 1;
            edge_list[edge_count++] = (Edge){u, v, weight};
        }
    }

    // Write to file
    FILE *file = fopen(filename, "w");
    if (!file) {
        fprintf(stderr, "Error opening file.\n");
        free(edge_list);
        exit(EXIT_FAILURE);
    }

    fprintf(file, "%lld %lld\n", vertices, edges);
    for (long long i = 0; i < edge_count; i++) {
        fprintf(file, "%lld %lld %d\n", edge_list[i].u, edge_list[i].v, edge_list[i].weight);
    }

    fclose(file);
    free(edge_list);
}

int main() {
    long long num_vertices, degree, base_edges, variability, num_edges;
    printf("Enter the number of vertices: ");
    scanf("%lld", &num_vertices);
    printf("Enter the desired average vertex degree: ");
    scanf("%lld", &degree);

    base_edges = num_vertices * degree / 2;
    variability = (long long)(base_edges * 0.05);
    num_edges = base_edges > variability ? base_edges - variability : num_vertices - 1;
    num_edges = num_edges < base_edges + variability ? num_edges : base_edges + variability;

    double actual_avg_degree = (2.0 * num_edges) / num_vertices;
    const char *output_file = "graph.txt";

    generate_weighted_graph(num_vertices, num_edges, output_file);
    printf("Graph of %lld nodes and %lld edges written to %s\n", num_vertices, num_edges, output_file);
    printf("Expected average degree: %lld, Actual average degree: %.2f\n", degree, actual_avg_degree);

    return 0;
}

File: Development/volume/graph_generator.py
import random

def generate_weighted_graph(vertices, edges, filename):
    if edges < vertices - 1:
        raise ValueError("Number of edges is too small to form a connected graph.")
    if edges > vertices * (vertices - 1) // 2:
        raise ValueError("Too many edges for the given number of vertices.")
    
    edge_set = set()
    
    # Step 1: Add edges to form a spanning tree (ensuring connectivity)
    for i in range(1, vertices):
        u = i
        v = random.randint(0, i - 1)  # Connect to a random previous node
        weight = random.randint(1, 100)
        edge_set.add((u, v, weight))
    
    # Step 2: Add remaining edges randomly
    while len(edge_set) < edges:
        u = random.randint(0, vertices - 1)
        v = random.randint(0, vertices - 1)
        if u != v:
            weight = random.randint(1, 100)
            edge_set.add((min(u, v), max(u, v), weight))
    
    # Write to the file
    with open(filename, 'w') as file:
        file.write(f"{vertices} {edges}\n")
        for u, v, weight in edge_set:
            file.write(f"{u} {v} {weight}\n")

if __name__ == "__main__":
    num_vertices = int(input("Enter the number of vertices: "))
    degree = int(input("Enter the desired average vertex degree: "))
    base_edges = num_vertices * degree // 2
    
    # Variability of ±5%
    variability = int(base_edges * 0.05)
    num_edges = max(base_edges - variability, num_vertices - 1)  # Ensure enough edges for connectivity
    num_edges = min(num_edges, base_edges + variability)
    
    # Calculate the actual average degree
    actual_avg_degree = (2 * num_edges) / num_vertices
    
    output_file = "graph.txt"

    try:
        generate_weighted_graph(num_vertices, num_edges, output_file)
        print(f"Graph of {num_vertices} nodes and {num_edges} edges written to {output_file}")
        print(f"Expected average degree: {degree}, Actual average degree: {actual_avg_degree:.2f}")
    except ValueError as e:
        print(f"Error: {e}")
